### PR TITLE
This PR adds support for packed GEMM on X86 and AARCH64 for FP32 and FP64

### PR DESCRIPTION
**Pull request by alheinecke**
Packed GEMM is operating in 3d data structures: Its a batch GEMM with the batch-dimension being the inner most dimension. The code vectorizes over this inner dimension: `C[N][M][B] += A[K][M][B] * B[N][K][B]`

### CODE REVIEW

File: src/generator_packed_gemm_aarch64.c [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+
+  /* load C accumulator */
+  libxsmm_generator_packed_gemm_aarch64_asimd_load_C( io_generated_code, i_gp_reg_mapping, i_micro_kernel_config, i_xgemm_desc, i_n_blocking, i_m_blocking, l_use_masking, i_packed_width );
+
+  /* k loop */
+  libxsmm_generator_loop_header_aarch64( io_generated_code, io_loop_label_tracker, i_gp_reg_mapping->gp_reg_kloop, i_xgemm_desc->k );
+
+  /* full vector load of A */
+  for ( l_m = 0; l_m < i_m_blocking; ++l_m ) {
+    if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:518]: Why do you check for masking? I think the routine here https://github.com/libxsmm/libxsmm/blob/main/src/generator_common_aarch64.c#L257 should also work with 0 masked elements.

File: src/generator_packed_gemm_aarch64.c [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+    libxsmm_aarch64_instruction_alu_compute_imm64( io_generated_code, LIBXSMM_AARCH64_INSTR_GP_META_ADD,
+                                                   i_gp_reg_mapping->gp_reg_a, i_gp_reg_mapping->gp_reg_help_0, i_gp_reg_mapping->gp_reg_a,
+                                                   (long long)i_packed_width*i_micro_kernel_config->datatype_size_in );
+  }
+
+  /* loop over the register block */
+  for ( l_n = 0; l_n < i_n_blocking; ++l_n ) {
+    /* load B */
+    if ( l_use_masking ) {
+      libxsmm_generator_vloadstore_masked_vreg_aarch64( io_generated_code, i_gp_reg_mapping->gp_reg_b, i_gp_reg_mapping->gp_reg_help_1,
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:536]: same as above

File: src/generator_packed_gemm_aarch64.c [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+    for ( l_m = 0; l_m < i_m_blocking; l_m++ ) {
+      if (0 != (LIBXSMM_GEMM_FLAG_BETA_0 & i_xgemm_desc->flags)) { /* Beta=0 */
+        libxsmm_aarch64_instruction_sve_compute( io_generated_code,
+                                                 LIBXSMM_AARCH64_INSTR_SVE_EOR_V,
+                                                 (l_n*i_m_blocking) + l_m, (l_n*i_m_blocking) + l_m, (unsigned char)-1, (l_n*i_m_blocking) + l_m,
+                                                 LIBXSMM_AARCH64_SVE_REG_UNDEF,
+                                                 LIBXSMM_AARCH64_SVE_TYPE_D );
+      } else {
+        if ( i_use_masking ) {
+          libxsmm_generator_vloadstore_masked_vreg_aarch64( io_generated_code, i_gp_reg_mapping->gp_reg_c, i_gp_reg_mapping->gp_reg_help_2,
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:604]: Same as above

File: src/generator_packed_gemm_aarch64.c [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+                                                                           const unsigned int                 i_use_masking,
+                                                                           const unsigned int                 i_packed_width ) {
+  unsigned int l_n = 0;
+  unsigned int l_m = 0;
+
+  /* store C accumulator */
+  for ( l_n = 0; l_n < i_n_blocking; l_n++ ) {
+    for ( l_m = 0; l_m < i_m_blocking; l_m++ ) {
+      if ( i_use_masking ) {
+        libxsmm_generator_vloadstore_masked_vreg_aarch64( io_generated_code, i_gp_reg_mapping->gp_reg_c, i_gp_reg_mapping->gp_reg_help_2,
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:647]: Same as above

File: src/generator_packed_gemm_aarch64.c [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+
+  /* load C accumulator */
+  libxsmm_generator_packed_gemm_aarch64_sve_load_C( io_generated_code, i_gp_reg_mapping, i_micro_kernel_config, i_xgemm_desc, i_n_blocking, i_m_blocking, l_use_masking, i_packed_width );
+
+  /* k loop */
+  libxsmm_generator_loop_header_aarch64( io_generated_code, io_loop_label_tracker, i_gp_reg_mapping->gp_reg_kloop, i_xgemm_desc->k );
+
+  /* full vector load of A */
+  for ( l_m = 0; l_m < i_m_blocking; ++l_m ) {
+    if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:696]: Same as above

File: src/generator_packed_gemm_aarch64.c [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+    }
+    libxsmm_aarch64_instruction_alu_compute_imm64( io_generated_code, LIBXSMM_AARCH64_INSTR_GP_META_ADD,
+                                                   i_gp_reg_mapping->gp_reg_a, i_gp_reg_mapping->gp_reg_help_0, i_gp_reg_mapping->gp_reg_a,
+                                                   (long long)i_packed_width*i_micro_kernel_config->datatype_size_in );
+  }
+
+  /* loop over the register block */
+  for ( l_n = 0; l_n < i_n_blocking; ++l_n ) {
+    /* load B */
+    if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_aarch64.c:713]: Same as above

File: src/generator_packed_gemm_avx_avx2_avx512.c [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+
+  for ( l_n = 0; l_n < i_n_blocking; l_n++ ) {
+    for ( l_m = 0; l_m < i_m_blocking; l_m++ ) {
+      if (0 != (LIBXSMM_GEMM_FLAG_BETA_0 & i_xgemm_desc->flags)) { /* Beta=0 */
+        libxsmm_x86_instruction_vec_compute_3reg( io_generated_code,
+                                                  i_micro_kernel_config->vxor_instruction,
+                                                  i_micro_kernel_config->vector_name,
+                                                  (l_n*i_m_blocking) + l_m, (l_n*i_m_blocking) + l_m, (l_n*i_m_blocking) + l_m );
+      } else {
+        /* in case of masking we need to distinguish between AVX/AVX2 and AVX512 */
```
Comment (egeor) [inline src/generator_packed_gemm_avx_avx2_avx512.c:446]: Wouldn't this work? https://github.com/libxsmm/libxsmm/blob/main/src/generator_common_x86.c#L124

File: src/generator_packed_gemm_avx_avx2_avx512.c [hunk 8]
```diff
@@ -0,0 +1,751 @@
...
+  if ( LIBXSMM_DATATYPE_F64 == LIBXSMM_GEMM_GETENUM_AB_COMMON_PREC( i_xgemm_desc->datatype ) ) {
+    l_avx_ac_move_instr = LIBXSMM_X86_INSTR_VMASKMOVPD;
+  } else {
+    l_avx_ac_move_instr = LIBXSMM_X86_INSTR_VMASKMOVPS;
+  }
+
+  for ( l_n = 0; l_n < i_n_blocking; l_n++ ) {
+    for ( l_m = 0; l_m < i_m_blocking; l_m++ ) {
+      /* in case of masking we need to distinguish between AVX/AVX2 and AVX512 */
+      if ( i_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_avx_avx2_avx512.c:502]: Same as above

File: src/generator_packed_gemm_avx_avx2_avx512.c [hunk 9]
```diff
@@ -0,0 +1,751 @@
...
+  /* load C accumulator */
+  libxsmm_generator_packed_gemm_avx_avx2_avx512_load_C( io_generated_code, i_gp_reg_mapping, i_micro_kernel_config, i_xgemm_desc, i_n_blocking, i_m_blocking, l_use_masking, i_packed_width );
+
+  /* k loop */
+  libxsmm_generator_gemm_header_kloop( io_generated_code, io_loop_label_tracker, i_gp_reg_mapping, i_micro_kernel_config, 0, 1 );
+
+  /* full vector load of A */
+  /* in case of masking we need to distinguish between AVX/AVX2 and AVX512 */
+  for ( l_m = 0; l_m < i_m_blocking; ++l_m ) {
+    if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_avx_avx2_avx512.c:596]: Same as above

File: src/generator_packed_gemm_avx_avx2_avx512.c [hunk 10]
```diff
@@ -0,0 +1,751 @@
...
+                                        l_m * i_packed_width * i_micro_kernel_config->datatype_size_in,
+                                        i_micro_kernel_config->vector_name,
+                                        i_n_blocking*i_m_blocking + l_m, 0, 1, 0 );
+    }
+  }
+
+  /* loop over the register block */
+  for ( l_n = 0; l_n < i_n_blocking; ++l_n ) {
+    if ( ( io_generated_code->arch >= LIBXSMM_X86_AVX512_VL128_SKX ) && ( io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) ) {
+      if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_avx_avx2_avx512.c:630]: Same as above

File: src/generator_packed_gemm_avx_avx2_avx512.c [hunk 11]
```diff
@@ -0,0 +1,751 @@
...
+      for ( l_m = 0; l_m < i_m_blocking; ++l_m ) {
+        libxsmm_x86_instruction_vec_compute_3reg( io_generated_code,
+                                                  i_micro_kernel_config->vmul_instruction,
+                                                  i_micro_kernel_config->vector_name,
+                                                  i_n_blocking*i_m_blocking + l_m,
+                                                  i_n_blocking*i_m_blocking + i_m_blocking,
+                                                  (l_n * i_m_blocking) + l_m );
+      }
+    } else if ( io_generated_code->arch >= LIBXSMM_X86_AVX2 ) {
+      if ( l_use_masking ) {
```
Comment (egeor) [inline src/generator_packed_gemm_avx_avx2_avx512.c:658]: Ditto

**Review by egeor — COMMENTED** [review]
Minor nit: It seems some of the "unified" vec moves (both x86 and aarch64) are not used. This would make the code much shorter/more readable/less error-prone

File: src/generator_packed_gemm_aarch64.c (hunk 1 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

File: src/generator_packed_gemm_aarch64.c (hunk 2 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

File: src/generator_packed_gemm_aarch64.c (hunk 4 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

File: src/generator_packed_gemm_aarch64.c (hunk 5 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

File: src/generator_packed_gemm_aarch64.c (hunk 6 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

File: src/generator_packed_gemm_avx_avx2_avx512.c (hunk 7 above)
Comment (alheinecke) [reply src/generator_packed_gemm_avx_avx2_avx512.c]: fixed

File: src/generator_packed_gemm_avx_avx2_avx512.c (hunk 8 above)
Comment (alheinecke) [reply src/generator_packed_gemm_avx_avx2_avx512.c]: fixed

File: src/generator_packed_gemm_avx_avx2_avx512.c (hunk 9 above)
Comment (alheinecke) [reply src/generator_packed_gemm_avx_avx2_avx512.c]: fixed

File: src/generator_packed_gemm_avx_avx2_avx512.c (hunk 10 above)
Comment (alheinecke) [reply src/generator_packed_gemm_avx_avx2_avx512.c]: fixed

File: src/generator_packed_gemm_avx_avx2_avx512.c (hunk 11 above)
Comment (alheinecke) [reply src/generator_packed_gemm_avx_avx2_avx512.c]: fixed

File: src/generator_packed_gemm_aarch64.c (hunk 3 above)
Comment (alheinecke) [reply src/generator_packed_gemm_aarch64.c]: fixed

**Comment by alheinecke** [comment]
thanks, good comment. unified and refactored the code.

**Comment by alheinecke** [comment]
@egeor : please give it a quick new review

**Pull request merged**